===================================================================
RCS file: /cvs/gcc/gcc/gcc/cp/ChangeLog,v
retrieving revision 1.3695
retrieving revision 1.3696
#diff -u -r1.3695 -r1.3696
#--- gcc/gcc/cp/ChangeLog	2003/09/19 10:00:14	1.3695
#+++ gcc/gcc/cp/ChangeLog	2003/09/19 14:07:56	1.3696
#@@ -1,3 +1,9 @@
#+2003-09-19  Kriang Lerdsuwanakij  <lerdsuwa@users.sourceforge.net>
#+
#+	PR c++/495
#+	* pt.c (tsubst_friend_class): Only use innermost template
#+	arguments for the injected friend class template.
#+
# 2003-09-19  Nathan Sidwell  <nathan@codesourcery.com>
# 
# 	PR c++/12332
===================================================================
RCS file: /cvs/gcc/gcc/gcc/cp/pt.c,v
retrieving revision 1.778
retrieving revision 1.779
diff -u -r1.778 -r1.779
--- gcc/gcc/cp/pt.c	2003/09/15 14:48:27	1.778
+++ gcc/gcc/cp/pt.c	2003/09/19 14:07:57	1.779
@@ -5001,6 +5001,8 @@
       DECL_USE_TEMPLATE (tmpl) = 0;
       DECL_TEMPLATE_INFO (tmpl) = NULL_TREE;
       CLASSTYPE_USE_TEMPLATE (TREE_TYPE (tmpl)) = 0;
+      CLASSTYPE_TI_ARGS (TREE_TYPE (tmpl))
+	= INNERMOST_TEMPLATE_ARGS (CLASSTYPE_TI_ARGS (TREE_TYPE (tmpl)));
 
       /* Inject this template into the global scope.  */
       friend_type = TREE_TYPE (pushdecl_top_level (tmpl));
===================================================================
RCS file: /cvs/gcc/gcc/gcc/testsuite/ChangeLog,v
retrieving revision 1.3067
retrieving revision 1.3068
#diff -u -r1.3067 -r1.3068
#--- gcc/gcc/testsuite/ChangeLog	2003/09/19 10:00:20	1.3067
#+++ gcc/gcc/testsuite/ChangeLog	2003/09/19 14:07:58	1.3068
#@@ -1,3 +1,8 @@
#+2003-09-19  Kriang Lerdsuwanakij  <lerdsuwa@users.sourceforge.net>
#+
#+	PR c++/495
#+	* g++.dg/template/friend24.C: New test.
#+
# 2003-09-19  Nathan Sidwell  <nathan@codesourcery.com>
# 
# 	PR c++/12332
/cvs/gcc/gcc/gcc/testsuite/g++.dg/template/friend24.C,v  -->  standard output
revision 1.1
--- gcc/gcc/testsuite/g++.dg/template/friend24.C
+++ -	2003-09-27 20:12:41.000000000 +0000
@@ -0,0 +1,27 @@
+// { dg-do compile }
+
+// Origin: Wolfgang Bangerth <bangerth@ticam.utexas.edu>
+
+// PR c++/495: Fail to locate primary class template that is
+// injected by friend declaration.
+
+template <int N> struct X
+{
+  template <int dim> friend struct Y;
+};
+
+X<2> x;
+
+template <int dim> struct Y
+{
+  void f (Y);
+  void g (Y);
+};
+
+template <int dim> void Y<dim>::f (Y)
+{
+}
+
+template <int dim> void Y<dim>::g (Y<dim>)
+{
+}
